File: py/ae/whocc/chains/web/js/table-page.js
import * as DIR from "./directories.js";
import * as MAPS from "./maps.js";

// ----------------------------------------------------------------------

function main() {
    console.log("table_page_data", table_page_data);
    make_title();
    const parts = table_page_data.parts.reduce((accumulator, elt) => {
        accumulator[elt.type] = [...(accumulator[elt.type] || []), elt];
        return accumulator;
    }, {});
    show_parts(parts, "individual", table_page_data.subtype_id);
    show_parts(parts, "chain", table_page_data.subtype_id);
    // for (let part_data of table_page_data.parts)
    //     show_part(part_data, table_page_data.subtype_id);
    // $("body").append(`<pre>\n${JSON.stringify(table_page_data, null, 2)}\n</pre>`);
}

// ----------------------------------------------------------------------

function show_parts(parts, key, subtype_id) {
    for (var part of parts[key]) {
        const part_title_text = part_title(part, subtype_id);
        if (part_title_text)
            $("body").append(`<hr>\n<h3><a href="chain?subtype_id=${subtype_id}&chain_id=${part.chain_id}" target="_blank">${part_title_text} ${part.date} (${part.chain_id})</a></h3>`);
        switch (key) {
        case "individual":
            show_individual_table_maps(parts);
            break;
        case "chain":
            show_chain_maps(part, parts);
            break;
        }
    }
}

// ----------------------------------------------------------------------

function show_individual_table_maps(data, chain_data) {
    const table = $("<table></table>").appendTo("body");
    const tr_title = $("<tr class='title'></tr>").appendTo(table);
    const tr = $("<tr class='image'></tr>").appendTo(table);
    data.individual.forEach((data_individual) => {
        table_page_data.coloring.forEach((coloring, coloring_no) => {
            MAPS.td_title_append(tr_title, tr, MAPS.map_td_with_title(data_individual, "individual", coloring, coloring_no === 0));
        });
    });
    // procrustes
    data.individual.forEach((data_individual) => {
        data.chain.forEach((data_chain) => {
            for (let merge_type of ["incremental", "scratch"]) {
                MAPS.td_title_append(tr_title, tr, MAPS.pc_td_with_title(data_individual, "individual", data_chain, merge_type, table_page_data.coloring[0]));
            }
        });
    });
    // TODO: grid test
    // TODO: serum correlation widget
    // TODO: table widget
}

// ----------------------------------------------------------------------

function show_chain_maps(data_chain, data) {
    const table = $("<table></table>").appendTo("body");
    table_page_data.coloring.forEach((coloring, coloring_no) => {
        const tr_title = $("<tr class='title'></tr>").appendTo(table);
        const tr = $("<tr class='image'></tr>").appendTo(table);
        for (let merge_type of ["incremental", "scratch"]) {
            MAPS.td_title_append(tr_title, tr, MAPS.map_td_with_title(data_chain, merge_type, coloring, coloring_no === 0));
        }
        MAPS.td_title_append(tr_title, tr, MAPS.pc_td_with_title(data_chain, "incremental", data_chain, "scratch", coloring)); // pc incremental vs. scratch
        if (data.individual && data.individual[0]) {
            for (let merge_type of ["incremental", "scratch"]) {
                MAPS.td_title_append(tr_title, tr, MAPS.pc_td_with_title(data_chain, merge_type, data.individual[0], "individual", coloring));
            }
        }
        if (data_chain.mcb) {
            MAPS.td_title_append(tr_title, tr, MAPS.map_td_with_title(data_chain, "mcb", coloring, coloring_no === 0));
            if (data.individual && data.individual[0]) {
                MAPS.td_title_append(tr_title, tr, MAPS.pc_td_with_title(data_chain, "mcb", data.individual[0], "individual", coloring));
            }
            for (let merge_type of ["incremental", "scratch"]) {
                MAPS.td_title_append(tr_title, tr, MAPS.pc_td_with_title(data_chain, merge_type, data_chain, "mcb", coloring));
            }
        }
        for (let merge_type of ["incremental", "scratch"]) {
            // TODO: grid test
        }
    });
}

// ----------------------------------------------------------------------

function part_title(part_data, subtype_id) {
    switch (part_data.type) {
    case "individual":
        return null;
    case "chain":
        switch (part_data.chain_id[0]) {
        case "f":
            return "Chain"; // `<a href="chain?subtype_id=${subtype_id}&chain_id=${part_data.chain_id}" target="_blank">Chain</a>`;
        case "b":
            return "Backward chain"; // `<a href="chain?subtype_id=${subtype_id}&chain_id=${part_data.chain_id}" target="_blank">Backward chain</a>`;
        default:
            return "? Chain"; // `<a href="chain?subtype_id=${subtype_id}&chain_id=${part_data.chain_id}" target="_blank">? Chain</a>`;
        }

    }
    return null;
}

// ----------------------------------------------------------------------

function make_title() {
    let prefix = `${DIR.lab_to_display[table_page_data.lab]} ${DIR.subtype_to_display[table_page_data.subtype]}`;
    if (table_page_data.subtype === "h3")
        prefix += ` ${DIR.assay_to_display[table_page_data.assay]} ${table_page_data.rbc || ""}`;
    $("#title").html(`${prefix} ${table_page_data.table_date}`);
}

// ----------------------------------------------------------------------

$(document).ready(() => main());

File: py/ae/whocc/chains/web/table_page.py
"""Table page: for one table date, collect the individual + chain maps involving it.

Ported from AD ``table_page.py``. Chart reads go through ``render.get_chart`` /
``render.chart_date`` (ae_backend), replacing AD's ``chart.get_chart`` (acmacs).
Paths are relative to cwd (the served root).
"""

import json
from pathlib import Path

from . import pages_util, render

# ======================================================================

sTablePage = """<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8" />
    {stylesheets}
    {remote_scripts}
    {inline_scripts}
    <title>{table_name}</title>
  </head>
  <body>
    <h2 id="title"></h2>
{body}
  </body>
</html>
"""

# ======================================================================


def table_page_html(ctx, subtype_id, table_date) -> str:
    remote_scripts = [
        "js/jquery.js",
        "js/table-page.js",
    ]
    stylesheets = [
        "js/maps.css",
        "js/table-page.css",
    ]
    data = collect_table_data(ctx, subtype_id=subtype_id, table_date=table_date)
    inline_scripts = [
        f"table_page_data =\n{json.dumps(data, separators=(',', ':'))};",
    ]
    return sTablePage.format(
        remote_scripts="\n    ".join(f'<script src="{script}" type="module"></script>' for script in remote_scripts),
        stylesheets="\n    ".join(f'<link rel="stylesheet" href="{stylesheet}">' for stylesheet in stylesheets),
        inline_scripts="\n    ".join(f'<script>\n{code}\n    </script>' for code in inline_scripts),
        table_name=f"{subtype_id} {table_date}",
        body="")

# ----------------------------------------------------------------------


def collect_table_data(ctx, subtype_id, table_date):

    def collect_table_data_part():
        for patt in ["i-*", "f-*", "b-*"]:
            for subdir in sorted(Path(subtype_id).glob(patt), reverse=True):
                entries = make_entries(subdir, patt[0] == "i")
                if entries:
                    yield {
                        "type": "individual" if patt[0] == "i" else "chain",
                        "chain_id": subdir.name,
                        **entries,
                    }

    def make_entries(subdir, individual):
        entries = {}
        # ignore temp files (with ~pid~ infix) made by export chart
        filenames = [fn for fn in subdir.glob(f"*{table_date}.*") if "~" not in str(fn)]
        if not filenames and subdir.name[0] == "b":  # backward-chain step-naming workaround
            filenames = list(subdir.glob(f"*.{table_date}-*"))
        for filename in filenames:
            key1, key2 = keys_for_filename(filename)
            if individual and key1 == "scratch":
                key1 = "individual"
            entries.setdefault(key1, {})[key2] = str(filename)
            if key2 == "ace":
                chart_date = render.chart_date(filename)
                entries[key1]["date"] = chart_date
                entries.setdefault("date", chart_date)
        return entries

    def keys_for_filename(filename):
        suffx = filename.suffixes
        if suffx[-1] == ".ace":
            if len(suffx) >= 2:                      # 123.20160113-20210625.incremental.ace
                return (suffx[-2][1:], "ace")
            return ("scratch", "ace")
        elif suffx[-1] == ".json":
            if len(suffx) >= 3:                      # 123.20160113-20210625.scratch.grid.json
                return (suffx[-3][1:], suffx[-2][1:])
            return ("scratch", suffx[-2][1:])
        return ("unknown", "".join(suffx))

    return {
        "subtype_id": subtype_id,
        "table_date": table_date,
        "parts": [part_data for part_data in collect_table_data_part() if part_data],
        **pages_util.format_subtype(ctx, subtype_id=subtype_id, date_range=[table_date, table_date]),
    }

# ======================================================================
